### ISSUE TITLE
Bug: pip package manager related commands raises error

### ISSUE THREAD

**Issue by PrashanthPuneriya** [anitab-org/mentorship-backend#664]
**Describe the bug**

`pip freeze`, `pip install` and other various pip related commands doesn't work and raises error due to the `six` older version package.
It raises `ImportError: cannot import name 'ensure_str' from 'six'`

**To Reproduce**

1. If you have a virtualenv folder then delete it and recreate the virtual environment as specified in the [docs](https://github.com/anitab-org/mentorship-backend#run-app)
2. Activate the virtual environment
3. Run `pip install -r requirements.txt` command.
4. Now to see the error run any `pip` related commands. For example run `pip freeze` command.
5. See the error

**Expected behavior**
`pip` related commands must be executed without running into any error.

**Screenshots**
![Screenshot from 2020-07-14 13-50-42](https://user-images.githubusercontent.com/42068010/87402468-3f5f3280-c5d9-11ea-9cc7-5bf97831eeba.png)

**Comment by PrashanthPuneriya** [comment]
I would like to take up and fix this bug.

**Comment by vj-codes** [comment]
Heyy @PrashanthPuneriya this issue does not seem to be valid as pip freeze does work with me .Can you see the Mentorship System API at local host on your computer?Might be a setup error from your side.I have approved the other issue!

cc : @anitab-org/mentorship-managers 
![Screenshot (18)](https://user-images.githubusercontent.com/60894542/87405303-e1cce500-c5dc-11ea-8fdb-0fdfc482cbb6.png)

**Comment by PrashanthPuneriya** [comment]
@vj-codes I think you haven't ran `pip install -r requirements.txt` command. Your virtual environment doesn't have all the requirements as specified in the requirements.txt file.

**Comment by vj-codes** [comment]
I did that when I first setup the repo .Can you run the project loacally? @PrashanthPuneriya

**Comment by PrashanthPuneriya** [comment]
@vj-codes Please follow the procedure to reproduce the bug. You need to recreate the virtual environment again.

**Comment by SanketDG** [comment]
@PrashanthPuneriya Last I checked, `mentorship-backend` does not work with 3.8. Were you able to run the application atleast once successfully?

**Comment by PrashanthPuneriya** [comment]
@vj-codes @SanketDG  No, I couldn't run the project locally. Not even once.

**Comment by SanketDG** [comment]
@PrashanthPuneriya I would say try with 3.7 then, if you want to set up the environment now.

**Comment by vj-codes** [comment]
Its working properly with me  for 3.8 on windows @SanketDG 
@PrashanthPuneriya You are on Ubuntu so dont have much idea, however I had faced issue #212 but the solution worked.

**Comment by PrashanthPuneriya** [comment]
@SanketDG @vj-codes I think I have fixed this issue and #212 . I would confirm my hypothesis and let you know.

**Comment by PrashanthPuneriya** [comment]
@vj-codes @SanketDG 
I am posting this comment as a new comer to this project. So, please review and give feedback. I am using `Python 3.8.2`.

## To fix #664 we got two work-arounds 

### Method-1 (Temporary work-around) :-

- Upgrade `six` to 1.14.0
- Upgrade `werkzeug` to >= 0.16.1

### Method-2 (Permanent work-around) :-

- #605 fixes this issue
- `flask-restplus` (which this project uses) created dependency issues for `six` and `wekzeug`. So, to fix this dependency issue we need to either follow Method-1 or Method-2(i.e. use `flask-restx`)
- You can read more about dependency issues from [here](https://github.com/pypa/virtualenv/issues/1551) and [here](https://github.com/noirbizarre/flask-restplus/issues/777)

## To fix #212

- Uninstall `cffi` (if this package isn't required. I haven't seen any dependency on this package until now. See [this comment](https://github.com/anitab-org/mentorship-backend/issues/664#issuecomment-658135641).)

**Comment by SanketDG** [comment]
#605 doesn't fix all the dependency bugs, it just upgrades to `flask-restx`.

> * Uninstall `cffi` (if this package isn't required. I haven't seen any dependency on this package until now). This apparently fixes #212

It's a pretty common library to be used by packages that are written in C and have python bindings, so it's probably installed as a dependency of some other package.

Can you do an audit of all packages in requirements to make sure no unneeded dependency is present? Then please open a PR with all dependencies updated to their latest versions! We are doing the same for [portal](https://github.com/anitab-org/portal)

Let me know if you have any questions.

**Comment by PrashanthPuneriya** [comment]
@SanketDG Okay! But, I think before auditing the dependencies for this project we must fix #400 (#400 removes major unrequired dependencies which makes the work much easier). 
Aren't we using python3 already? If we are using python3 then we must move from virtualenv to venv i.e. fix #400.

**Comment by PrashanthPuneriya** [comment]
@SanketDG @vj-codes @isabelcosta I did some dig up in commits to find out when `cffi` package was first included in the `requirements.txt` file. #207 was the pull request when it was first included.
I haven't seen any code in that particular pull request which are dependent on the following packages:- `cffi`, `asn1crypto`, `cryptography` and these packages are not futher required by other packages.

So, I think we can safely remove `cffi` package and go ahead with the temporary work-around as mentioned [above](https://github.com/anitab-org/mentorship-backend/issues/664#issuecomment-658105326).

@SanketDG this might be the first step towards auditing the dependencies.

**Comment by vj-codes** [comment]
@SanketDG  so this issue is valid right?
If yes then, 
@PrashanthPuneriya you can only work for one issue at a time. So, which one of the two issues would you like to fix first?

**Comment by vj-codes** [comment]
Remember that the other issue is for first timers only so you can't work on that later @PrashanthPuneriya

**Comment by PrashanthPuneriya** [comment]
@vj-codes I would like to fix #663 issue first

**Comment by PrashanthPuneriya** [comment]
### Update
@vj-codes @SanketDG @anitab-org/mentorship-managers I have updated the [above comment](https://github.com/anitab-org/mentorship-backend/issues/664#issuecomment-658105326). Please go through it once again and review it.

@isabelcosta If you review this [comment](https://github.com/anitab-org/mentorship-backend/issues/664#issuecomment-658135641) then we can proceed futher.

Thanks!

**Comment by SanketDG** [comment]
@PrashanthPuneriya Feel free to do any of the above, what ultimately matters is that we upgrade each and every dependency and then automate it using https://github.com/marketplace/dependabot-preview

Please don't selectively upgrade, upgrade every dependency and see what breaks. We can iterate from there.

And yes, you can choose to not upgrade anything involves `flask-restplus`. since my PR already covers that.

**Comment by SanketDG** [comment]
@PrashanthPuneriya Can you update the description to explicitly mention what needs to be done here?

**Comment by PrashanthPuneriya** [comment]
@SanketDG Just tested this. When did we merged #605 :P?. #605 solved most of the problem. After merging #713 we can close this issue.

**Comment by SanketDG** [comment]
Yeah that's what I thought, sounds good, we can remove the OSH label right?

**Comment by PrashanthPuneriya** [comment]
I have tested this only on Python 3.8. Let me confirm with other versions and decide whether to close this issue or not :)

**Comment by Abjcodes** [comment]
I would like to work on this issue as a part of OSH. Can you assign me this issue? Thank You.

**Comment by PrashanthPuneriya** [comment]
Sure @Abjcodes ! But, can you wait for a while until #713 gets merged? As of now this issue stands valid but, we need to test again after merging the mentioned PR and see for other new issues.

**Comment by soumyjain14** [comment]
Hi, I would like to work in the open source Hack, can someone assign any task to me?

**Comment by codesankalp** [comment]
@SanketDG @PrashanthPuneriya I want to work on this issue and I think I can fix this easily.

**Comment by PrashanthPuneriya** [comment]
@SanketDG This issue is still valid. Earlier, I tested it without even running `pip related command`. My bad :laughing:

**Comment by PrashanthPuneriya** [comment]
I am making this issue as first-timers-only since, it is quite simple to fix.
Sorry @Abjcodes, this means I can't assign you this issue since, you have already created a PR.

**Comment by PrashanthPuneriya** [comment]
Assigning @soumyjain14 :)

**Comment by soumyjain14** [comment]
Hey thanks, @PrashanthPuneriya , Let me try this

**Comment by soumyjain14** [comment]
sorry unable to find a clue :(
New to open source, don't know how to start

**Comment by PrashanthPuneriya** [comment]
No worries @soumyjain14! Did you successfully set up the project? At which part are you clueless? Let me help you getting started 😃

**Comment by PrashanthPuneriya** [comment]
If you have successfully set up the project locally and replicated the error then start by googling out the error and always read the prevs comments to get more hints :)

**Comment by PrashanthPuneriya** [comment]
Are you still working on this @soumyjain14 ? Or should I assign someone else?

**Comment by PrashanthPuneriya** [comment]
@pallavisavant @Anmollenka Do anyone of you want to work on this issue? It's a first-timer-only issue.

**Comment by pallavisavant** [comment]
Yes i would like to work...lemme take this up

**Comment by PrashanthPuneriya** [comment]
Assigning you @pallavisavant :)

**Comment by PrashanthPuneriya** [comment]
Oh sorry @pallavisavant I can't assign you. This is a first-timer issue. My bad I didn't checked thoroughly. I will mention you whenver I see a suitable issue.

**Comment by pallavisavant** [comment]
@PrashanthPuneriya oh okay ya thanks...👍

**Comment by Anmollenka** [comment]
@PrashanthPuneriya Can you please inform me if there any issues with outreach as the label.I am searching them but I am not able to find it.Thank You.

**Comment by devkapilbansal** [comment]
@PrashanthPuneriya I want to work on this issue

**Comment by PrashanthPuneriya** [comment]
Sure @KapilBansal320. Assigning you :)

**Comment by devkapilbansal** [comment]
Hi, @PrashanthPuneriya I tested it on Python 3.8.2 on Ubuntu 20.04 and it seems to work perfectly for me. 
**TLDR**
Here are some images confirming the same
![s1](https://user-images.githubusercontent.com/42766576/93340859-09ddfc00-f84b-11ea-802e-88af5fd6ce50.png)
![s2](https://user-images.githubusercontent.com/42766576/93340861-09ddfc00-f84b-11ea-9380-fca1765b592b.png)
![s3](https://user-images.githubusercontent.com/42766576/93340867-0b0f2900-f84b-11ea-9607-354cb3034c1a.png)
![s4](https://user-images.githubusercontent.com/42766576/93340878-0cd8ec80-f84b-11ea-863f-1566495f6f09.png)

Although I will test it on Windows too but I don't think this issue is still valid or can you please guide me where I am going wrong?

**Comment by PrashanthPuneriya** [comment]
Hey @KapilBansal320 I just tested it again by recreating the environment and I am still getting the same error.
![image](https://user-images.githubusercontent.com/42068010/93421217-5f0f2180-f8ce-11ea-9853-608d4035c278.png)

**Comment by devkapilbansal** [comment]
> Hey @KapilBansal320 I just tested it again by recreating the environment and I am still getting the same error.
> ![image](https://user-images.githubusercontent.com/42068010/93421217-5f0f2180-f8ce-11ea-9853-608d4035c278.png)

@PrashanthPuneriya Okay can you provide me with the OS you are using also the version of Python too

**Comment by PrashanthPuneriya** [comment]
@KapilBansal320 
OS -> Ubuntu 20.04
Python 3.8.2
pip 20.0.2

**Comment by devkapilbansal** [comment]
@PrashanthPuneriya please review pull request #877
